### DISCUSSION TITLE
Polymorphic identity refer to joined table.

### DISCUSSION THREAD

**Discussion by matchratings** [sqlalchemy/sqlalchemy#7310] [Usage Questions] [answered]
New to everything - OOP, python, sqlalchemy etc, so this very well may not be a sqlalchemy question - I will take any guidance! Thanks in advance. 

I want to subclass ( create Cupmatch(Matches))) using a type column that is not an attribute of the superclass, but of a related class (League). Not sure what the strategy is to execute this.

```
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy import Column, Integer, String,ForeignKey
from sqlalchemy.orm import relationship
Base = declarative_base()


class League(Base):
    __tablename__='league'
    leagueid = Column(Integer, primary_key=True)
    type=Column(String(20))

class Leagueseason(Base):
    __tablename__='leagueseason'
    leagueseasonid= Column(Integer,primary_key=True)
    leagueid=Column(Integer, ForeignKey(League.leagueid))
    league=relationship("League", foreign_keys=[leagueid], backref='leagueseason')

class Matches(Base):
    __tablename__='matches'
    matchid= Column(Integer,primary_key=True)
    leagueseasonid = Column(Integer,ForeignKey(Leagueseason.leagueseasonid))

 # I want this polymorphic identity to refer to match.leagueseason.league.type 
class Cupmatch(Matches):
    __mapper_args__ = {'polymorphic_identity':'cup'
    }
```

**Comment by zzzeek** [comment, marked as answer]
we have some examples for this pattern, which is not directly built in but can be worked out, at https://github.com/sqlalchemy/sqlalchemy/wiki/PolymorphicOnAnotherTable .  I've just updated the code for modern use and made sure both examples run successfully.

**Reply by zzzeek** [reply]
there's other ways to do this too that are more fancy, but all of them require having some kind of lookup table of AType objects at some point, at least on the persistence side.  there's kind of no way around that happening somewhere, either up front, or in event hooks, etc.

**Reply by matchratings** [reply]
Thanks I read the example pattern per your recommendation. I will try to adapt, but my example is slightly more complicated because the polymorphic identity is not in an adjacent table, but two tables away (two joins). I am new to column_property, select etc. Should I make a query using column property/select with two joins? I know how to do this in sql but not yet in sqlalchemy.

**Reply by matchratings** [reply]
I think I answered my question on multiple joins. This code seems to do the trick:
```
    type_name = column_property(
      select(League.type).select_from(join(Leagueseason,League)).where(Leagueseason.leagueseasonid==leagueseasonid).correlate_except(League,Leagueseason).scalar_subquery())
```

**Reply by zzzeek** [reply]
if that works, that's good, but for scaling up I'd look at how efficient that query is from an EXPLAIN perspective (new to SQL also?  I'm talking about [this](https://dev.mysql.com/doc/refman/8.0/en/explain.html), there are similar commands for all the non-sqlite databases) .  the approach that populates polymorphic_map up front is a little uglier but more efficient from the SQL side of things.

this does come up from time to time so maybe someday I will build in a more comprehensive feature for this.

**Reply by matchratings** [reply]
Sorry for the delay - got very distracted. Yes I'm new to sql as well, and efficiency, is not my priority right this second, but obviously something I would like to have a better feel for long term. The sql explain command looks useful. I have the gist of polymorphic_map from google, but can't find documentation on anything resembling the pattern for "the approach that populates polymorphic_map up front". Do you have a link by any chance? Thanks again for all your time - extremely helpful.

**Reply by zzzeek** [reply]
the approach that populates polymorphic map up front is the example I linked at https://github.com/sqlalchemy/sqlalchemy/wiki/PolymorphicOnAnotherTable , the second code example.
